src/components/editableTable/GroupEditableTableRow.js

```diff
@@ -0,0 +1,266 @@
+import TableCell from '@material-ui/core/TableCell';
+import EditIcon from '@material-ui/icons/Edit';
+import Save from '@material-ui/icons/Save';
+import Clear from '@material-ui/icons/Clear';
+import Delete from '@material-ui/icons/Delete';
+import Button from '@material-ui/core/Button';
+import Tooltip from '@material-ui/core/Tooltip';
+import { green, red } from '@material-ui/core/colors';
+import { makeStyles } from '@material-ui/styles';
+import EditableTableCell from './EditableTableCell';
+import columnsProps from './columnsProps';
+
+const useStyles = makeStyles(theme => ({
+    button: {
+        maxWidth: theme.spacing(5),
+        minWidth: theme.spacing(5),
+        maxHeight: theme.spacing(3),
+        minHeight: theme.spacing(3),
+        padding: 0
+    },
+    row: {
+        backgroundColor: props => {
+            if (props.toBeSaved) {
+                return green[50];
+            }
+            if (props.toBeDeleted) {
+                return red[50];
+            }
+            return 'transparent';
+        }
+    }
+}));
+
+export default function GroupEditableTableRow({
+    row,
+    columns,
+    editable,
+    updateRow,
+    validateRow,
+    removeRow,
+    isRowValid,
+    resetRow,
+    deleteRowPreEdit,
+    handleEditClick,
+    setRowToBeSaved,
+    setRowToBeDeleted,
+    removeRowOnDelete,
+    ...rest
+}) {
+    const [rowValid, setRowValid] = useState();
+
+    const isRowValidRef = useRef();
+
+    const classes = useStyles({ toBeSaved: row.toBeSaved, toBeDeleted: row.toBeDeleted });
+
+    useEffect(() => {
+        isRowValidRef.current = isRowValid;
+    }, [isRowValid]);
+
+    useEffect(() => {
+        if (row.editing) {
+            let valid = true;
+
+            if (validateRow) {
+                valid = validateRow(row);
+            }
+
+            columns.forEach(column => {
+                if (column.required === true && !row[column.id]) {
+                    valid = false;
+                }
+            });
+
+            if (isRowValidRef.current && row.id) {
+                isRowValidRef.current(valid, row.id);
+            }
+
+            setRowValid(valid);
+        }
+    }, [validateRow, columns, row]);
+
+    const handleCancelClick = () => {
+        if (row.isNewRow) {
+            removeRow(row.id);
+        } else {
+            resetRow(row);
+        }
+    };
+
+    const handleValueChange = (propertyName, newValue) => {
+        updateRow(row, null, propertyName, newValue);
+    };
+
+    const handleEditButtonClick = () => {
+        handleEditClick(row.id, true);
+    };
+
+    const handleDeleteClick = () => {
+        if (removeRowOnDelete || row.isNewRow) {
+            removeRow(row.id);
+        } else {
+            setRowToBeDeleted(row.id, true);
+        }
+    };
+
+    const handleSaveClick = () => {
+        setRowToBeSaved(row.id, true);
+    };
+
+    const saveButtonEnabled = () => {
+        return !!(rowValid && !row.toBeDeleted && !row.toBeSaved);
```

Comment:
is !! doing anything here?  Doesn't the thing in the brackets evaluates to a boolean anyway?